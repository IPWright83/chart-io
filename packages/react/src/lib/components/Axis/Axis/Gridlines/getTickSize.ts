import type { IMargin, IPosition } from "@chart-it/types";
import { logAndThrowError } from "../../../../utils";

/**
 * Obtain the tick size used for gridlines
 * @param  position     The position of the axis [left, right, top, bottom]
 * @param  width        The width for the chart
 * @param  height       The height for the chart
 * @param  margin       The margin object for the chart
 * @return              The transform
 */
const getTickSize = (position: IPosition, width: number, height: number, margin: IMargin): number => {
    if (width === 0 || height === 0) {
        return 0;
    }

    switch (position) {
        case "left":
        case "right":
            return width - margin.left - margin.right;
        case "top":
        case "bottom":
            return height - margin.top - margin.bottom;
        default:
            logAndThrowError("E002", `Invalid position (${position}) was provided to the Axis`);
    }
};

export { getTickSize };
