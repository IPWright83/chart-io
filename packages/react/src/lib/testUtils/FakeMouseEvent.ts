export interface IFakeMouseEventValues {
    pageX?: number;
    pageY?: number;
    offsetX?: number;
    offsetY?: number;
    x?: number;
    y?: number;
}

/**
 * Create a Fake MouseEvent for testing
 */
export class FakeMouseEvent extends MouseEvent {
    /**
     * Construct a new instance of a FakeMoustEvent
     * @param  type        The type of event, e.g. "click", "mouseenter"
     * @param  values      An object containing values for the x,y co-ordinates for the event
     */
    constructor(type: string, values: IFakeMouseEventValues) {
        const { pageX, pageY, offsetX, offsetY, x, y, ...mouseValues } = values;
        super(type, mouseValues);

        Object.assign(this, {
            offsetX: offsetX || 0,
            offsetY: offsetY || 0,
            pageX: pageX || 0,
            pageY: pageY || 0,
            x: x || 0,
            y: y || 0,
        });
    }
}
