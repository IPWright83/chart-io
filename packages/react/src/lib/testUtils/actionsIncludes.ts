/**
 * Assert that all the expected dispatch calls to the Redux store were seen
 * @param {string[]} dispatchCalls The actual calls
 * @param {string[]} expectedCalls The expected calls
 */
export function actionsIncludes(dispatchCalls: string[], expectedCalls: string[]) {
    expectedCalls.forEach((expected) => {
        expect(dispatchCalls).toContain(expected);
        dispatchCalls.splice(dispatchCalls.indexOf(expected), 1);
    });
}
