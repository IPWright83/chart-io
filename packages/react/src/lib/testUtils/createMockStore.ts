import { defaultChartState, defaultEventState, IStore } from "../store";

/**
 * Creates a mock Redux store for testing with
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStore(state: { chart?: any; event?: any }): IStore {
    return {
        getState: () => ({
            chart: {
                ...{ ...defaultChartState, id: "test" },
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        dispatch: jest.fn(),
        subscribe: jest.fn(),
    } as unknown as IStore;
}

/**
 * Creates a mock Redux store for Storybook
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStorybookStore(state: { chart?: any; event?: any }) {
    return {
        getState: () => ({
            chart: {
                ...defaultChartState,
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        dispatch: () => {},
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        subscribe: () => {},
    } as unknown as IStore;
}
