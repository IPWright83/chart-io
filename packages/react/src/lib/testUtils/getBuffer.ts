/**
 * Returns the image buffer for a Canvas
 * @param canvas    The HTML canvas that we're going to extract the buffer from
 * @returns         The image buffer
 */
export function getBuffer(canvas: HTMLCanvasElement) {
    const image = canvas.toDataURL("image/png");
    const data = image.replace(/^data:image\/\w+;base64,/, "");
    // eslint-disable-next-line no-undef
    return Buffer.from(data, "base64");
}
