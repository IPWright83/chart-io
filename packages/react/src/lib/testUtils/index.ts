export * from "./actionsIncludes";
export * from "./FakeMouseEvent";
export * from "./createMockStore";
export * from "./getBuffer";
export * from "./wait";
export * from "./renderChart";
export * from "./testEvents";
