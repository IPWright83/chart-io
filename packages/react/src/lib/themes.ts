import type { IColor, ITheme } from "@chart-io/types";

const light: ITheme = {
    axis: {
        stroke: "#333333" as IColor,
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    droplines: {
        strokeDasharray: 4,
        strokeOpacity: 1,
        strokeWidth: 2,
    },

    markers: {
        size: 6,
        stroke: "white" as IColor,
        strokeWidth: 2,
        shadow: false,
    },

    font: {
        family: "sans-serif",
        size: 12,
    },

    background: "#ffffff" as IColor,

    crosshair: {
        stroke: "#000000" as IColor,
        strokeOpacity: 0.5,
        strokeWidth: 1,
        strokeDasharray: 0,
    },

    gridlines: {
        stroke: "#d3d3d3" as IColor,
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    tooltip: {
        background: "#ffffff" as IColor,
        text: "#000000" as IColor,
        opacity: 0.8,
        padding: 8,
    },

    legend: {
        background: "#ffffff" as IColor,
        text: "#000000" as IColor,
        opacity: 1,
        padding: 8,
        border: "#ccc",
        defaultMaxWidth: 200,
        defaultMaxHeight: 50,
    },

    series: {
        opacity: 0.7,
        selectedOpacity: 1,
        colors: [
            "#99c1dc",
            "#fc998e",
            "#fdc381",
            "#c2e587",
            "#fff18c",
            "#c999ca",
            "#a4dcd2",
            "#d6efd1",
            "#ffffc2",
            "#cbc8e1",
            "#fdd7ea",
            "#e1e1e1",
        ] as IColor[],
    },
};

const dark: ITheme = {
    axis: {
        stroke: "#cccccc" as IColor,
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    droplines: {
        strokeDasharray: 4,
        strokeOpacity: 1,
        strokeWidth: 2,
    },

    background: "#333333" as IColor,

    font: {
        family: "sans-serif",
        size: 12,
    },

    crosshair: {
        stroke: "#cccccc" as IColor,
        strokeOpacity: 0.8,
        strokeWidth: 1,
        strokeDasharray: 0,
    },

    markers: {
        size: 6,
        stroke: "white" as IColor,
        strokeWidth: 2,
        shadow: false,
    },

    gridlines: {
        stroke: "#555555" as IColor,
        strokeOpacity: 0.8,
        strokeWidth: 1,
    },

    tooltip: {
        background: "#333333" as IColor,
        text: "#ffffff" as IColor,
        opacity: 0.8,
        padding: 8,
    },

    legend: {
        background: "#333333" as IColor,
        text: "#ffffff" as IColor,
        opacity: 1,
        padding: 8,
        border: "#ccc",
        defaultMaxWidth: 200,
        defaultMaxHeight: 50,
    },

    series: {
        opacity: 1,
        selectedOpacity: 0.7,
        colors: [
            "#51cbdf",
            "#ee709d",
            "#f8c86c",
            "#677be7",
            "#90e5a4",
            "#91564b",
            "#ec73c1",
            "#7f7f7f",
            "#bdbf2f",
            "#00becf",
        ] as IColor[],
    },
};

const themes = {
    light,
    dark,
};

export { themes };
