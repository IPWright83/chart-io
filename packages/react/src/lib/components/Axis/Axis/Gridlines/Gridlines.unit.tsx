import * as d3 from "@chart-it/d3";
import { Provider } from "react-redux";
import React from "react";
import { render } from "@testing-library/react";

import { createMockStore } from "../../../../testUtils";
import { themes } from "../../../../themes";

import { getTickSize } from "./getTickSize";
import { Gridlines } from ".";

describe("Gridlines", () => {
    const width = 1000;
    const height = 500;
    const margin = { left: 10, right: 20, top: 30, bottom: 40 };

    const store = createMockStore({
        chart: {
            theme: themes.light,
            dimensions: {
                width,
                height,
                plotMargin: margin,
            },
        },
    });

    describe("getTickSize", () => {
        const plotHeight = height - margin.top - margin.bottom;
        const plotWidth = width - margin.left - margin.right;

        describe("should return full width", () => {
            it("for left axis", () => {
                expect(getTickSize("left", plotWidth, plotHeight)).toBe(970);
            });
            it("for right axis", () => {
                expect(getTickSize("right", plotWidth, plotHeight)).toBe(970);
            });
        });

        describe("should return 0", () => {
            it("for 0 width", () => {
                expect(getTickSize("right", 0, plotHeight)).toBe(0);
            });
            it("for 0 height", () => {
                expect(getTickSize("right", plotWidth, 0)).toBe(0);
            });
        });

        describe("should return full height", () => {
            it("for top axis", () => {
                expect(getTickSize("top", plotWidth, plotHeight)).toBe(430);
            });
            it("for bottom axis", () => {
                expect(getTickSize("bottom", plotWidth, plotHeight)).toBe(430);
            });
        });

        it("throws an error with an invalid position", () => {
            // @ts-expect-error: Testing out the runtime validation
            expect(() => getTickSize("invalid", plotWidth, plotHeight, margin)).toThrow();
        });
    });

    describe("component", () => {
        it("renders horizontal gridlines", async () => {
            const scale = d3
                .scaleLinear()
                .domain([0, 100])
                .range([0, width - margin.left - margin.right]);

            const { asFragment } = render(
                <Provider store={store}>
                    <svg>
                        <Gridlines position="bottom" scale={scale} />
                    </svg>
                </Provider>
            );

            expect(asFragment()).toMatchSnapshot();
        });

        it("renders vertical gridlines", async () => {
            const scale = d3
                .scaleLinear()
                .domain([0, 10])
                .range([0, height - margin.top - margin.bottom]);

            const { asFragment } = render(
                <Provider store={store}>
                    <svg>
                        <Gridlines position="left" scale={scale} />
                    </svg>
                </Provider>
            );

            expect(asFragment()).toMatchSnapshot();
        });
    });
});
