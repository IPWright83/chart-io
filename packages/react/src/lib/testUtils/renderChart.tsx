import type { IData, IScale } from "@chart-io/types";
import { Provider } from "react-redux";
import React from "react";
import { render } from "@testing-library/react";

import { createMockStore } from "./createMockStore";
import { wait } from "./wait";

import { themes } from "../themes";

export async function renderChart({
    children,
    data,
    scales = {},
    store,
}: {
    children: JSX.Element;
    data?: IData;
    scales?: Record<string, IScale>;
    store?: any;
}) {
    const mockStore = createMockStore({
        chart: {
            animationDuration: 0,
            dimensions: {
                height: 200,
                width: 200,
            },
            data,
            scales: Object.entries(scales).reduce((result, [key, scaleFunc]) => {
                result[key] = {
                    domain: scaleFunc.domain(),
                    range: scaleFunc.range(),
                    scale: scaleFunc,
                };

                return result;
            }, {}),
            theme: themes.light,
        },
    });

    const { asFragment, container } = render(
        <Provider store={store ?? mockStore}>
            <svg>{children}</svg>
        </Provider>
    );

    await wait(10);

    return { asFragment, container, store: store ?? mockStore };
}
