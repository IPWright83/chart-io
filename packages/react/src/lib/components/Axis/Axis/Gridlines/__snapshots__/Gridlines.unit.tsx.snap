// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gridlines component renders horizontal gridlines 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-gridlines bottom"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      style="color: rgb(211, 211, 211); stroke-opacity: 0.8; stroke-width: 1; pointer-events: none;"
      text-anchor="middle"
    >
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(97.5,0)"
      >
        <line
          stroke="currentColor"
          y2="-430"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(194.5,0)"
      >
        <line
          stroke="currentColor"
          y2="-430"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(291.5,0)"
      >
        <line
          stroke="currentColor"
          y2="-430"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(388.5,0)"
      >
        <line
          stroke="currentColor"
          y2="-430"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(485.5,0)"
      >
        <line
          stroke="currentColor"
          y2="-430"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(582.5,0)"
      >
        <line
          stroke="currentColor"
          y2="-430"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(679.5,0)"
      >
        <line
          stroke="currentColor"
          y2="-430"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(776.5,0)"
      >
        <line
          stroke="currentColor"
          y2="-430"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(873.5,0)"
      >
        <line
          stroke="currentColor"
          y2="-430"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(970.5,0)"
      >
        <line
          stroke="currentColor"
          y2="-430"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="3"
        />
      </g>
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Gridlines component renders vertical gridlines 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-gridlines left"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      style="color: rgb(211, 211, 211); stroke-opacity: 0.8; stroke-width: 1; pointer-events: none;"
      text-anchor="end"
    >
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(0,43.5)"
      >
        <line
          stroke="currentColor"
          x2="970"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(0,86.5)"
      >
        <line
          stroke="currentColor"
          x2="970"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(0,129.5)"
      >
        <line
          stroke="currentColor"
          x2="970"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(0,172.5)"
      >
        <line
          stroke="currentColor"
          x2="970"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(0,215.5)"
      >
        <line
          stroke="currentColor"
          x2="970"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(0,258.5)"
      >
        <line
          stroke="currentColor"
          x2="970"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(0,301.5)"
      >
        <line
          stroke="currentColor"
          x2="970"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(0,344.5)"
      >
        <line
          stroke="currentColor"
          x2="970"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(0,387.5)"
      >
        <line
          stroke="currentColor"
          x2="970"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        />
      </g>
      <g
        class="tick"
        opacity="0.000001"
        transform="translate(0,430.5)"
      >
        <line
          stroke="currentColor"
          x2="970"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        />
      </g>
    </g>
  </svg>
</DocumentFragment>
`;
